hazelcast/include/hazelcast/client/IAtomicLong.h

```diff
@@ -111,8 +126,127 @@ namespace hazelcast {
 
             virtual void destroy();
 
+            /**
+             * Atomically adds the given value to the current value.
+             * <p>
+             * This method will dispatch a request and return immediately an
+             * {@link ICompletableFuture}.
+             * <p>
+             * The operations result can be obtained in a blocking way, or a callback
+             * can be provided for execution upon completion, as demonstrated in the
+             * following examples:
+             * <pre><code>
+             * boost::shared_ptr<ICompletableFuture<int64_t> > future = atomicLong.addAndGetAsync(13);
+             * // do something else, then read the result
+             *
+             * // this method will block until the result is available
+             * int64_t result = future.get();
+             * </code></pre>
+             * <pre><code>
+             *   future->andThen(boost::shared_ptr<ExecutionCallback<V> >(new  MyExecutionCallback()));
+             * </code></pre>
+             *
+             * @param delta the value to add
+             * @return an {@link ICompletableFuture} bearing the response
+             * @since cluster version 3.7
+             */
+            boost::shared_ptr<ICompletableFuture<int64_t> > addAndGetAsync(int64_t delta);
+
+            /**
+             * Atomically sets the value to the given updated value
+             * only if the current value {@code ==} the expected value.
+             * <p>
+             * This method will dispatch a request and return immediately an
+             * {@link ICompletableFuture}.
+             *
+             * @param expect the expected value
+             * @param update the new value
+             * @return an {@link ICompletableFuture} with value {@code true} if successful;
+             * or {@code false} if the actual value was not equal to the expected value
+             * @since cluster version 3.7
+             */
+            virtual boost::shared_ptr<ICompletableFuture<bool> > compareAndSetAsync(int64_t expect, int64_t update);
```

Comment:
Why most of these new methods are virtual? As far as I understand IAtomicLong is not further inherited.

Reply:
They are virtual methods from the impl::AtomicLongInterface.

Reply:
All methods are from `AtomicLongInterface` but only new ones are virtual. 
As far as I know, if we make these virtual when a user calls these methods over IAtomicLong then there will be unnecessary virtual table lookup.  And when I read the code, it tells me that IAtomicLong will be further inherited which is not the case. I am not really concerned about performance impact since I am not sure if it causes a visible penalty, but in terms of readability of the code, it gives the wrong impression.

Reply:
Removed the virtual keywords from the method signatures.